### ISSUE TITLE
Stop grad for certain inputs

### ISSUE THREAD

**Issue by rgreenblatt** [google/jax#9630]
When taking multiple derivatives wrt. different inputs, it can be useful to stop gradients with respect to specific input(s). For instance, consider the following:
```python
import jax


def stop_grad_for_input(x, inp):
    # div by inp happens to work in this specific case as it inverts the effect
    # of inp while still being the identity, but I'd like to do this in
    # general.
    return x / inp


def f(a_0, a_1):
    x = 4 * a_0 * a_1
    return stop_grad_for_input(3 * x, inp=a_0) * stop_grad_for_input(2 * x, inp=a_1)


print(jax.jacfwd(lambda a_1: jax.jacfwd(lambda a_0: f(a_0, a_1))(1.0))(1.0))  # should be (4 * 3) * (2 * 4) = 96
```

I'd like to be able to do this in general, but I'm not sure how to do so.
I'm interested in implementing this, so pointers on where to get started would be helpful. (My current best guess would be to manipulate the trace in a `defjvp` function somehow).

**Comment by rgreenblatt** [comment]
Thanks for the response.
As far as the reactions go, I shared the link with some coworkers and they upvoted it.

I ended up coming up with a approach which works for what I need (stopping gradients for a particular jvp trace level), but is some cursed/sketchy:
```python
from typing import FrozenSet, Iterable
from copy import copy

import jax
from jax.interpreters.ad import JVPTracer

from aann.tools.grad_modify_query import remove_jvp_trace_level_for_inps


def remove_jvp_trace_levels(x, levels: FrozenSet[int]):
    if not isinstance(x, JVPTracer) or len(levels) == 0:
        return x

    if x._trace.level in levels:
        return remove_jvp_trace_levels(x.primal, levels.difference([x._trace.level]))
    else:
        x = copy(x)
        x.primal = remove_jvp_trace_levels(x.primal, levels)
        x.tangent = copy(x.tangent)
        x.tangent.val = remove_jvp_trace_levels(x.tangent.val, levels)
        return x


def remove_jvp_trace_level_for_inps(x, inps: Iterable[JVPTracer]):
    return remove_jvp_trace_levels(x, frozenset({inp._trace.level for inp in inps}))


def test_removes():
    def f(a_0: JVPTracer, a_1: JVPTracer, a_2: JVPTracer, a_3: JVPTracer):
        x = 4 * a_0 + 3 * a_1
        y = 13 * x * a_2 + 23 * x * a_3

        return ((3 * remove_jvp_trace_level_for_inps(y, [a_0, a_3])) + 100_000) * (
            (2 * remove_jvp_trace_level_for_inps(y, [a_1, a_2])) + 100_000
        )

    def get():
        return jax.jacfwd(
            lambda a_3: jax.jacfwd(
                lambda a_2: jax.jacfwd(lambda a_1: jax.jacfwd(lambda a_0: f(a_0, a_1, a_2, a_3))(1.0))(1.0)
            )(1.0)
        )(1.0)

    assert get() == (3.0 * 13 * 3) * (4 * 23 * 2)
    assert jax.jit(get)() == (3.0 * 13 * 3) * (4 * 23 * 2)


if __name__ == "__main__":
    test_removes()
```
I'll probably end up sticking with this approach as it's more convenient than using a wrapper.

**Comment by mattjj** [comment]
Glad to hear you found something which works for you! It may be useful to keep in mind that the complexity of autodiff is exponential in the order, so calling `jacfwd` for each argument like that may get slow for compile times and/or execution times. (Conceptually, if you're familiar with dual numbers and how they're related to forward-mode AD, it's because the number of max-degree-one monomials in R[x1, ..., xk] is 2^k. That may be extremely cryptic unless you're an AD expert though.)
